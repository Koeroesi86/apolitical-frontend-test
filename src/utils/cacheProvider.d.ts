declare const _default: {
    get: (key: any) => Promise<any>;
    set: (key: any, value: any) => Promise<void>;
};
export default _default;
