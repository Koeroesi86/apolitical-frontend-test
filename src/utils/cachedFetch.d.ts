declare const _default: (url: any) => Promise<any>;
export default _default;
