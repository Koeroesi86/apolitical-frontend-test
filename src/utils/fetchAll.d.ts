declare const fetchAll: <T>(url: any) => Promise<T[]>;
export default fetchAll;
