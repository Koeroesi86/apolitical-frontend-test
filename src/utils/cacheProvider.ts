export default {
  get: async (key) => {
    const stored = sessionStorage.getItem(`url-${encodeURIComponent(key)}`);
    return stored ? JSON.parse(stored) : undefined;
  },
  set: async (key, value) => {
    sessionStorage.setItem(`url-${encodeURIComponent(key)}`, JSON.stringify(value));
  }
}