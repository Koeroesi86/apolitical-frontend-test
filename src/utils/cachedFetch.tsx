import axios from 'axios';
import cacheProvider from './cacheProvider';

export default async (url) => {
  const cached = await cacheProvider.get(url);
  if (cached !== undefined) return cached;

  const data = await axios({url, method: 'GET'}).then(res => res.data);
  await cacheProvider.set(url, data);

  return data;
}