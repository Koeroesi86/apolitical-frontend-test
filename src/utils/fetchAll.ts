import PageData from '../types/PageData';
import cachedFetch from './cachedFetch';

const fetchAll = async <T>(url): Promise<T[]> => {
  const data: PageData<T> = await cachedFetch(url);

  if (typeof data.next === 'string') {
    return [
      ...data.results,
      ...await fetchAll<T>(data.next)
    ] as T[];
  }

  return data.results as T[];
}

export default fetchAll;
