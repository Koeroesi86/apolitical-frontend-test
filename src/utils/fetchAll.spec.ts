import fetchAll from './fetchAll';
import cachedFetch from './cachedFetch';
import {mocked} from 'ts-jest/utils';

const testUrl = 'https://example.com/1';
jest.mock('./cachedFetch', () => ({
  __esModule: true,
  default: jest.fn(async (url: string) => {
    if (url === 'https://example.com/2') {
      return {
        previous: testUrl,
        results: [
          '3', '4'
        ]
      };
    }

    return {
      next: 'https://example.com/2',
      results: [
        '1', '2'
      ]
    };
  }),
}));

describe('fetchAll', () => {
  const mockedCachedFetch = mocked(cachedFetch, true);
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should fetch next pages', async () => {
    const result = await fetchAll(testUrl);

    expect(cachedFetch).toHaveBeenCalledTimes(2);
    expect(result).toMatchSnapshot('result');
    expect(mockedCachedFetch.mock.calls).toMatchSnapshot('cachedFetch calls');
  });
});
