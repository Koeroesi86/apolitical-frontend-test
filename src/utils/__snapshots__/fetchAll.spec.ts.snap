// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchAll should fetch next pages: cachedFetch calls 1`] = `
Array [
  Array [
    "https://example.com/1",
  ],
  Array [
    "https://example.com/2",
  ],
]
`;

exports[`fetchAll should fetch next pages: result 1`] = `
Array [
  "1",
  "2",
  "3",
  "4",
]
`;
